gnat2goto/driver/tree_walk.adb

```diff
@@ -4796,6 +4796,30 @@ package body Tree_Walk is
                        Range_Check     => False));
       end if;
 
+      if Nkind (N) = N_Subprogram_Body and then
+        Present (Corresponding_Spec (N)) and then
+        No_Return_Functions_Set.Contains
+          (Intern (Unique_Name (Corresponding_Spec (N))))
+      then
+         declare
+            No_Return_Check_Symbol : constant Irep := Symbol_Expr
+              (Get_Ada_Check_Symbol
+                 (Name           => "__CPROVER_Ada_Pragma_No_Return",
+                  A_Symbol_Table => Global_Symbol_Table,
+                  Source_Loc     => Internal_Source_Location));
+            No_Return_Check_Args : constant Irep := Make_Argument_List;
+            No_Return_Check_Call : constant Irep := Make_Code_Function_Call
+              (Arguments       => No_Return_Check_Args,
+               I_Function      => No_Return_Check_Symbol,
+               Lhs             => Make_Nil (Internal_Source_Location),
+               Source_Location => Internal_Source_Location,
+               I_Type          => Make_Void_Type);
+         begin
+            Append_Argument (No_Return_Check_Args, Get_Int32_T_Zero);
+            Append_Op (Reps, No_Return_Check_Call);
```

Comment:
This seems to add an assert at the end of a block - how does this handle returns inside of a block, like e.g.

```
procedure Test is
  procedure Does_Not_Return (X : Integer) is
  begin
    loop
      if X < 10 then
        return;
      end if;
    end loop;
  end Does_Not_Return;
  pragma No_Return (Does_Not_Return);
begin
  Does_Not_Return (5);
end Test;
```
?

Reply:
That's unfortunate.

Reply:
Yes; this is what I was getting at when I said : 

_One thing we should be careful of; I don't know if GNAT does any checking of the pragma at a syntactic level. If not then one could attach this to a function or procedure which contains return statements and so need to check that all of these are unreachable._

My apologies for not being clearer.